feat: multi system

Karrq opened the pull request:
Hello! I'm a mac user and wanted to make use of your derivations to manage my solana install. 
Unfortunately, it seemed like there was no support for mac at the time, but looking at the forks it seems like @0xPlish was working on a [multi-system derivation](https://github.com/0xPlish/solana-nix/tree/multi-system), also making use of flakes.

For me it @0xPlish's work wasn't building properly, in my case `crossbeam-epoch` (for `solana-cli`) wasn't able to build due to `no_atomic.rs` being a symlink to a file outside the crate root (in the repo root), thus the source wasn't available for `buildRustPackage`. 
I tried multiple ways to override the dependendy but couldn't figure it out, so in the end I switched to `crane` which handled the build out of the box.

Then I had an issue with `anchor-cli`, where the unpatched sources were used to build the dependencies of the binary (as part of a caching step, due to crane's build methodology) and thus the resulting binary had erroneous behaviour.

I also excluded the stripping of rlib files since rustc was unable to compile libcore (among others) without those.

Finally, I expanded the `systemMapping` in `solana-platform-tools` to also hold the output hashes.

0xPlish committed:
add flake; make derivations multi-system

0xPlish committed:
add flake

0xPlish committed:
Use oxalica overlay; add packages from local

0xPlish committed:
formatting

Karrq committed:
chore: formatting

Karrq committed:
feat: migrate to crane

fix: solana-cli build w/ crossbeam-epoch

Karrq committed:
fix(anchor-cli): patch before dep crate is built

Karrq committed:
fix(anchor): pre-patch sources

Karrq committed:
fix(solana): restore appended flags

Karrq committed:
fix(platform-tools): avoid stripping rlib files

Karrq committed:
fix(platform-tools): support all target systems

0xPlish commented:
Amazing @Karrq looking forward to trying your version. I didn't quite get my branch working.

Karrq committed:
feat(anchor-cli): support v0.30.1

Karrq committed:
feat(platform-tools): multiversion

deps(platform-tools): partial 1.43 support

Karrq committed:
feat(anchor): use appropriate platform-tools ver

fix(anchor): avoid calling +nightly on bundled cargo

Karrq committed:
refactor(anchor-cli): versionsDeps

fix(flake): don't package anchor 0.30.1 directly
fix(anchor): 0.30.1 full patch
feat: passthru.otherVersions

arijoon commented on criterion.nix:
```diff
@@ -1,9 +1,4 @@
-, pkg-config
-, criterion
-}:
+{ lib, fetchFromGitHub, cmake, pkg-config, criterion }:
```
which formatter are you using? (not sure if the official one collapses back into a single line)

Karrq replied:
nixfmt v0.6.0.
I just ran it across the repo and indeed everything was collapsed.

Do you use a different formatter? Could you let me know which one so I can format everything (or at least what I touched) with it

arijoon replied:
think I used `nixfmt nixpkgs-unstable-2024-12-04`
You can also use the latest rfc https://discourse.nixos.org/t/nix-formatting-team-full-nixpkgs-reformat/61867

Karrq replied:
Done with `nixfmt-tree`!

arijoon commented on flake.nix:
```diff
@@ -0,0 +1,59 @@
+{
+  description = "Solana development setup with Nix.";
+  inputs = {
+    nixpkgs.url = "github:nixos/nixpkgs/c5e2528c7c4ec05ce05a563e3be64f3525b278ad";
```
do we need to re-import these as flakes or can you simply use `pkgs.nix` here and only use flake-parts to create different systems? The reason is that it could cause a discrepancy between using flakes and non flake setup and the version are already fixed with niv sources so can simply import those inside this file (given there's no overriding of nixpkgs within additional sources, importing through flake inputs does not bring an additional advantage)

Karrq replied:
I will try importing it from nix/sources.nix, sounds like a good idea. I will also double check that we have compatibility with non-flakes.
If that proves to be non-flake friendly I will instead switch this input to the same one as the niv sources, and make use of `flake-compat` to export a non-flake derivation, how does that sound?

arijoon replied:
sounds good, you can also import the pkgs.nix and pass it the system override for it to make the pkgs attrset

Karrq replied:
Done with `nixfmt-tree`!

Karrq replied:
I decided to unify the approach by:
* using the same revs as specified in the niv sources for the flake inputs
* using flake-compat to define pkgs.nix, thus having it consistent with the flake

arijoon commented on flake.nix:
```diff
@@ -0,0 +1,59 @@
+  }:
+    flake-parts.lib.mkFlake {inherit inputs;} {
+      systems = [
+        "x86_64-linux"
+        "aarch64-linux"
+        "x86_64-darwin"
+        "aarch64-darwin"
+        "x86_64-windows"
```
do we know if this'll actually work on all these systems?

Karrq replied:
Added CI workflow!

arijoon commented on pkgs.nix:
```diff
@@ -1,23 +1,21 @@
-    }
-    )
+      solana-source = self.callPackage (import ./solana-source.nix) {};
+      solana-platform-tools =
+        self.callPackage (import ./solana-platform-tools.nix) {};
```
in flake.nix, there's a custom crane override for these. Is that not needed here?

Karrq replied:
`crane` is just a way for nix to orchestrate the build for rust crates, but in this case platform-tools is just a set of binaries as we don't built them from source in the derivation, hence why it's not needed.

arijoon commented on shell.nix:
```diff
@@ -1,13 +1,4 @@
-  ];
+let pkgs = import ./pkgs.nix { };
+in pkgs.mkShell {
+  buildInputs = with pkgs; [ anchor-cli solana-cli solana-rust yarn nodejs ];
```
formatter is collapsing into a single line again, this is undesired

arijoon reviewed (commented):
Thanks @Karrq for the PR and the work. This is a clear improvement to support multi systems and allow flake access so its easier to get started with a single command

Unfortunately I don't really have the expertise to maintain this due to unfamiliarity with `crane` and not having access to a Mac system to know if it'll break at a future point in time

If you'd like to maintain this as a fork, I'd be happy to link to your fork in the README file as superseding this repo for mac/multi system instead of just x86 linux.
Alternatively we'll need to setup some ci/github actions to make sure it'll remain functional across all systems

Karrq commented:
> Alternatively we'll need to setup some ci/github actions to make sure it'll remain functional across all systems

I'll investigate this, I agree that would be ideal, and also allow the project to ensure it works in other platforms without having to test it manually

Karrq committed:
fix(platform-tools): 1.43 hashes

style(anchor-cli): formatting

Karrq committed:
style(tree): nixfmt-tree

Karrq committed:
deps: match inputs to locked from niv

Karrq committed:
feat: flake-compat

Karrq committed:
feat(flake): export overlay

Karrq committed:
refactor: use flake-compat for pkgs.nix

chore: cleanup niv sources

Karrq committed:
style: remove unnecessary imports

Karrq committed:
feat(ci): initial workflow

Karrq commented:
Alright, I formatted the codebase with the rfc formatter, unified all inputs (to be managed by flakes), made sure that the current interface (shell, pkgs) is supported, and added a CI workflow that should check that the tools at least build properly.

I think to actually see the workflow running I'd need to add it to `main`, as I can't even run it as "workflow dispatch" on this branch. If you want I could open a separate PR with CI for the existing methodology, so this PR would also update that workflow

arijoon reviewed (commented):
Thanks we can merge to master and see it in action, its mainly for  the future updates to make sure other systems don't break. Just need to change the workflows to run on `master` instead of `main`

Karrq committed:
fix(ci): run against `master`

Karrq commented:
Ok, I will keep an eye out and make a followup PR with eventual fixes if necessary. 
Fixed the branch name 🙈

arijoon commented:
Unfortunately the current system does not work due to this https://github.com/solana-foundation/anchor/pull/3663

arijoon commented:
I've tried to address it here: https://github.com/arijoon/solana-nix/commit/47468968245d661fa1dfd1a919de0e85e5384ec9

I managed to get a working build on Linux, given the toolchain hasn't changed, I'm hoping it works on mac too, but I couldn't find a way without giving anchor the nightly build. Really wish rust projects would stop depending on unstable features, this seems like an abuse of nightly channel when its meant to try new features, not include them in stable software
